tests/sccache_cargo.rs

```diff
@@ -3,53 +3,101 @@
 //! Any copyright is dedicated to the Public Domain.
 //! http://creativecommons.org/publicdomain/zero/1.0/
 
-#![deny(rust_2018_idioms)]
+use anyhow::{Context, Result};
+use once_cell::sync::Lazy;
+
+use assert_cmd::prelude::*;
+use chrono::Local;
+use predicates::prelude::*;
+use std::ffi::OsString;
+use std::fs;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
 
-#[cfg(all(not(target_os = "windows"), not(target_os = "macos")))]
 #[macro_use]
 extern crate log;
 
-/// Test that building a simple Rust crate with cargo using sccache results in a cache hit
-/// when built a second time.
-#[test]
-#[cfg(all(not(target_os = "windows"), not(target_os = "macos")))]
-fn test_rust_cargo() {
-    test_rust_cargo_cmd("check", &[]);
-    test_rust_cargo_cmd("build", &[]);
+static SCCACHE_BIN: Lazy<PathBuf> = Lazy::new(|| assert_cmd::cargo::cargo_bin("sccache"));
+static CARGO: Lazy<OsString> = Lazy::new(|| std::env::var_os("CARGO").unwrap());
+static CRATE_DIR: Lazy<PathBuf> =
+    Lazy::new(|| Path::new(file!()).parent().unwrap().join("test-crate"));
```

Comment:
The documentation for `assert_cmd::cargo::cargo_bin` recommends to cache the result for performance reason. I'm not quite sure how up to date this is, but the issue linked from the documentation mentioned something like up to 0.1-0.3 seconds per evaluation.

For the other ones I guess it is not so important and I could just replace the lazy static with a function.

Reply:
> The documentation for assert_cmd::cargo::cargo_bin recommends to cache the result for performance reason.

Good dig, nice work.
FWIW, I'm assuming that that performance reason only matters if the variable is sometimes unused. Since we're guaranteed to need `SCCACHE_BIN` for this test, we're guaranteed to be running the `Lazy` closure, and therefore it can be inlined I believe.

Reply:
> FWIW, I'm assuming that that performance reason only matters if the variable is sometimes unused.

No, this is not the reason. once_cell::Lazy, is initialized on first access (hence lazy), but on any further accesses the value is already available and the closure is not executed. So the performance benefit this is used for is if you are evaluating something multiple times and know that the result can be cached.

Reply:
I guess it would be more clear if I used the full path for the type annotation, since only `Lazy` is somewhat ambiguous.

Reply:
> No, this is not the reason. once_cell::Lazy, is initialized on first access (hence lazy), but on any further accesses the value is already available and the closure is not executed

Right, but without `Lazy`, it's equivalent to the closure work (in our case, the code `assert_cmd::cargo::cargo_bin("sccache")`) being run immediately at runtime. Either way, the value is re-used. By removing `Lazy`, the only functional change that occurs is that the "closure work" happens earlier, IIUC.

So, the only way that `Lazy` would be associated with a performance benefit is if, due to conditional logic, the variable is never accessed, so the closure work is never redundantly resolved.
Since we always access the variable in our tests, this performance benefit is never encountered, right?

Reply:
> Right, but without Lazy, it's equivalent to the closure work being run immediately at runtime

I'm not sure I follow. Do you mean something like
```
 static CARGO: String = std::env::var("CARGO").unwrap(); 
```

That doesn't work because
```
error[E0015]: calls in statics are limited to constant functions, tuple structs and tuple variants
```

So I'm not sure if I'm missing something, but I don't see how to avoid using a lazily initialized variable for a global that can't be initialized at compile time.

Reply:
Oh, I see, I was imagining something like:
```suggestion
const SCCACHE_BIN: PathBuf = assert_cmd::cargo::cargo_bin("sccache");
const CARGO: OsString = std::env::var_os("CARGO").unwrap();
const CRATE_DIR: PathBuf = Path::new(file!()).parent().unwrap().join("test-crate");
```

But, it looks like quite a few of these functions aren't `const`-able yet.
Ok, let's carry on with `Lazy`, my mistake :)